pykafka/producer.py

```diff
@@ -111,15 +158,112 @@ def __repr__(self):
             id_=hex(id(self))
         )
 
-    def _send_request(self, broker, req, attempt):
+    def __enter__(self):
+        """Context manager entry point - start the producer"""
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Context manager exit point - stop the producer
+
+        If __exit__ is being called as a result of self._worker_exception
+        being raised, don't wait on the workers to finish. If __exit__ is being
+        called because the with: block has ended, do wait for workers to finish.
+        """
+        # If the thread crashed, don't wait for it
+        self.stop(wait=self._worker_exception is None)
+
+    @raise_worker_exceptions
+    def start(self):
+        """Set up data structures and start worker threads"""
+        if not self._running:
+            self._owned_brokers = {}
+            for partition in self._topic.partitions.values():
+                if partition.leader.id not in self._owned_brokers:
+                    self._owned_brokers[partition.leader.id] = OwnedBroker(
+                        self, partition.leader)
+            self._running = True
+
+    @raise_worker_exceptions
+    def stop(self, wait=True):
+        """Mark the producer as stopped
+
+        :param wait: Whether to wait for all inflight messages to be sent
+            before returning
+        :type wait: bool
+        """
+        self._running = False
+        if wait:
+            self._wait_all()
+
+    @raise_worker_exceptions
+    def produce(self, messages):
+        """Produce a set of messages.
+
+        :param messages: The messages to produce
+        :type messages: Iterable of str or (str, str) tuples
+        """
+        def _partition_messages():
+            """Assign messages to partitions using the partitioner."""
+            partitions = self._topic.partitions.values()
+            for message in messages:
+                if isinstance(message, basestring):
+                    key = None
+                    value = message
+                else:
+                    key, value = message
+                yield (key, str(value)), self._partitioner(partitions, message).id
+        if not self._running:
+            raise ProducerStoppedException()
+        pairs = itertools.izip(self._produce(_partition_messages()), messages)
+        for message_enqueued in pairs:
+            yield message_enqueued
```

Comment:
After discussion, we agree that it makes more sense to stop supporting sequences of messages in `produce`. That will make the whole `block_on_queue_full` issue go away. It will also require that we use a major version bump.